airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py

```diff
@@ -261,7 +284,14 @@ def get_stream_state(self) -> Optional[Mapping[str, StreamState]]:
             }
         }
         """
-        return copy.deepcopy(self._parent_state)
```

Comment:
I'm not sure I get this change: why would we return only the state for one partition when setting the parent_state? Why does it matter if it is the last partition being processed or not? 

I'm a bit afraid of that because this clearly indicates that the SubstreamPartitionRouter must know about PerPartition states which makes it a circular logical dependency. This is dangerous because now, every time someone modifies one, they will need to know that there is an impact in the other one even though this is not explicit.

Reply:
I discussed a bit with @brianjlai and there are a couple of points which would make us want to remove the `last` from this interface:
* This is only needed by `CartesianPartitionRouter`
* `CartesianPartitionRouter` is basically not used (there are custom connectors that re-implements it but they could be replaced by something like PerPartition
* Having parent states with `CartesianPartitionRouter` can gives undetermined state and I would prefer not to support that. For example, assuming state {cursor_field: 10} for parent 1 (whatever parent 1 is) and state {cursor_field: 20} for parent 2, those one of the two streams will overwrite the cursor value of the other and this will probably lead to data loss

Honestly, I think the concept of `CartesianPartitionRouter` should be deprecated and eventually removed. For those reasons, I prefer to keep the interface clean (i.e. without the `last` parameter) and ignore the CartesianPartitionRouter's problems for now. We should just flag this class as deprecated.

Reply:
I disagree. The Cartesian product is a basic concept — the need to combine multiple lists into a product of elements. For example, when adding filters to Jira issues, projects and filters must be combined as a product.

Reply:
After a sync with @tolik0, here is my understanding of the situation: 
* There are cases where this is used which I've missed ([example](https://github.com/airbytehq/airbyte/blob/3203dc09878a3b70ea43392b65c1d1cf7105456d/airbyte-integrations/connectors/source-asana/source_asana/manifest.yaml#L329-L336)). However, we think this case is scoffed and expect it to not work as the dev expected it to work
* There are some hypothetical cases which @tolik0 identified with filters. For example, let's get all the comments with upvotes and downvotes (filters) for all the issues (parent stream). That being said, we don't have a prod use case in our pool of connectors for now.
* The worst thing we've identified about changing the interface is that now, all the callers of get_stream_state needs to track which partition is last. This is annoying because we would need to transpose this logic in the concurrent framework as well.

So as a middle ground, we've decided that:
* We won't change the `get_stream_state` interface because of the added complexity and low value
* We will not delete `CartesianPartitionRouter` as we see potential cases where this could be useful in the future
* Because we don't know how things will evolve and we want to avoid scope creep for that, we will simply add a log saying that the state management will not work when the CartesianProductSlicer is instantiated and don't do anything in the methods that relates to state. The reasoning is that is was scoffed anyway right now and we don't want to maintain it with the `last` boolean param